### ISSUE TITLE
A failed 'test' op does not prevent the patch from applying

### ISSUE THREAD

**Issue by biofractal** [winduptoy/mongoose-json-patch#1]
I am having trouble getting my 'test' operations to be respected. According to the spec a sequence of patches should be atomic:

_From RFC-6902_

>If a normative requirement is violated by a JSON Patch document, or if an operation is not successful, evaluation of the JSON Patch document SHOULD terminate and application of the entire patch document SHALL NOT be deemed successful.

This does not appear to be the case as the following test illustrates:

**original document**
```
{ _id: 54f8469f8cf9510a8a2c5edd,                                                           
  stock: 889,                                         
  __v: 0 }    
 ```
 
**patches**
```
[ { op: 'test', path: '/stock', value: 890 }, 
  { op: 'replace', path: '/stock', value: 888 } ]  
```

**result of apply()**
```
true
```                   
**error generated**
```
null                                        
```                                          
**final patched document**
```
{ _id: 54f8469f8cf9510a8a2c5edd,                                                           
  stock: 888,                                         
  __v: 0 }
```

I think the problem is probably to be found in the [fast-json-patch](https://github.com/Starcounter-Jack/JSON-Patch) package where there is currently the following open issue:

- [apply() only returns result of last operation](https://github.com/Starcounter-Jack/JSON-Patch/issues/64)

However even if fast-json-patch did return `false`correctly for a failed sequence of patches you do not process the result. Perhaps you, like me, imagined that should an op 'test' fail then the patches would not be applied and the document would remain unchanged?

**Comment by winduptoy** [comment]
@biofractal I've got it in a try/catch around the `apply()` statement, so I think I assumed that any failures would throw an error, and thus not save the document.

Is the expected behavior of [fast-json-patch](https://github.com/Starcounter-Jack/JSON-Patch) to return false and **not** throw an error when a test returns false?

Do you think something like this will do it?

```javascript
var result;

try {
	result = jsonpatch.apply(this, patches, true);
} catch(err) {
	return callback(err);
}

if(!result) {
	return callback(new Error('Something went wrong with the patch'));
}

//Save the document (or parent document if this happens to be a subdocument)
if(this.ownerDocument && this.ownerDocument()) return this.ownerDocument().save(callback);
return this.save(callback);
```

If you send a PR I'd be happy to pull it.

**Comment by winduptoy** [comment]
Fixed and merged in #2. Thanks, @biofractal.
